### ISSUE TITLE
How to select submatrix while keeping the dimensions intact

### ISSUE THREAD

**Issue by szarnyasg** [GraphBLAS/LAGraph#83]
I am trying to extract the induced subgraph of a set of vertices from a graph. In matrix terms, these vertices can be captured as a set of indexes and the extraction operation would be selecting these rows & columns from the matrix, which can be achieved with `GrB_Matrix_extract`. However, I'd like to perform the operation in a way that the dimension of the matrix does not change (so that the vertex indexes are kept, which can be used for subsequent 'join' operations).

I have to ideas to perform this:

1. Create a `diagm(vertex ids)` matrix and multiply the adjacency matrix from both left and right. Quick demo code in Julia:

    ```julia
    using LinearAlgebra

    #    1 2 3 4 5 6
    A = [0 0 1 1 0 0 #1
         0 0 1 1 0 0 #2
         1 1 0 0 1 0 #3
         1 1 0 0 0 1 #4
         0 0 1 0 0 0 #5
         0 0 0 1 0 0]#6

    S = diagm([0, 1, 0, 1, 0, 1])

    S*A*S
    ```

    This returns:
    ```
    julia> S = diagm([0, 1, 0, 1, 0, 1])
    6×6 Array{Int64,2}:
    0  0  0  0  0  0
    0  1  0  0  0  0
    0  0  0  0  0  0
    0  0  0  1  0  0
    0  0  0  0  0  0
    0  0  0  0  0  1

    julia> S*A*S
    6×6 Array{Int64,2}:
    0  0  0  0  0  0
    0  0  0  1  0  0
    0  0  0  0  0  0
    0  1  0  0  0  1
    0  0  0  0  0  0
    0  0  0  1  0  0
    ```

2. Use `GxB_select` and pass the set of indices as the `thunk`, then use a custom function to check whether both `i` and `j` are in the set of vertex identifiers. A custom `GxB_select` example is available in Borůvka's algorithm:
https://github.com/GraphBLAS/LAGraph/blob/18c709693a4f7ebd186d7cff7bb586f4af465a4a/Source/Algorithm/LAGraph_cc_boruvka.c#L58-L62

Are there any other ways of doing this? What would be the recommended approach in terms of expected performance?

(An obvious tradeoff is that #1 works with any GraphBLAS implementation. In our current work with @marci543 we are using SuiteSparse:GraphBLAS so using #2 is also possible.)

**Comment by DrTimothyAldenDavis** [comment]
I think the option of multiplying by the 2 diagonal matrices is the best
one, because it's more portable.
However, the GxB_select option will likely be faster.  Hard to say how much
though, without trying it out.

**Comment by szarnyasg** [comment]
Thanks - I've drafted an initial prototype of the two approaches: https://github.com/szarnyasg/LAGraph/blob/extract-submatrix-with-indices/Source/Algorithm/LAGraph_inducedsubgraph.c

Ultimately, I believe this function could make its way to LAGraph but it will need a bit of tuning & testing first.

### Implementation

Here's the signature of the function:

```c
GrB_Info LAGraph_inducedsubgraph // compute the subgraph induced by vertices V
                                 // in A
(
    GrB_Matrix *Chandle,         // output matrix
    const GrB_Matrix A,          // input matrix
    const GrB_Index *V,          // sorted list of vertex indices
    GrB_Index nv,                // number of vertex indices
    bool use_select              // use GxB_select (faster but not portable)
)
```

The approach to use by the function can be chosen with the `use_select` parameter.
* Implementing the diagonal approach for the `use_select == false` option was fairly straightforward.
* The `GxB_select` variant used for the `true` option was more tricky as it necessitated the creation of a new `GrB_Type` called `V_ind_type` which is a struct storing the indices. This is wrapped into a `GxB_Scalar`, the thunk, and is passed to the `GxB_select` operation which uses a custom selection operation `sel_op`. This uses binary search (`bsearch`) to determine whether indices `i` and `j` are in the provided set of vertex indices `V`.

### Performance

Here's how it performs on my laptop with an i7-8550u CPU:

I did a microbenchmark using the datagen-7_5-fb Graphalytics data set's BOOL variant (~600k vertices, 17M(×2) undirected edges). My first test filtered for every other vertex, namely, the ones indexed with an even number:

Approach #1 (multiplying with the diagonal):
```console
Time elapsed:       2.66 seconds, 1 threads
Time elapsed:       1.82 seconds, 2 threads
Time elapsed:       1.01 seconds, 4 threads
Time elapsed:       0.99 seconds, 8 threads
```

Approach #2, using `GxB_select`:
```console
Time elapsed:      19.99 seconds, 1 threads
Time elapsed:       8.56 seconds, 2 threads
Time elapsed:       6.23 seconds, 4 threads
Time elapsed:       5.47 seconds, 8 threads
```

In this test, approach #1 clearly wins.

In my second test, I selected every 100th vertex, in which case #1 won by an even larger margin: <0.1 seconds vs. 5-20 seconds.

In these very rudimentary tests, the diagonal variant was faster. Maybe I made some suboptimal choices when implementing [`use_select == true`](https://github.com/szarnyasg/LAGraph/blob/extract-submatrix-with-indices/Source/Algorithm/LAGraph_inducedsubgraph.c#L105-L123) -- please let me know if you have any feedback.

### Naming

Regarding the naming of this function, I think `inducedsubgraph` is not a great name as it's basically the dual of `extract_matrix`. Maybe something like `extract_matrix_with_original_indices` would work better.

Let me know if you have any comments & suggestions.

**Comment by DrTimothyAldenDavis** [comment]
Looks great.  One comment though.  I would consider a dense vector to
scatter the pattern of V into, and then the bsearch would be an O(1)-time
lookup.  I would calloc the array and then set it to true where the indices
appear in V. That would take O(n) time to initialize, however, which is not
good for the hypersparse case.

The mxm solution would work easily if n = 2^60, just as long as nv is much
smaller of course.  The dense vector option for GxB_select would fail if it
tried to allocate O(n) space.

Even so, I'm not sure if GxB_select would be faster than mxm.

Have you tried swapping the order of the two mxm?  That can affect the
performance.  Assuming A is held by row, the order you have should be the
fastest one.  If A is held by column, the order of the two mxm's could be
swapped to speed it up.  But that would require a test with GxB_get to
query the GrB_Matrix format, which is not portable.

**Comment by szarnyasg** [comment]
Thanks for the detailed feedback!

I added two modes of operation, that's chosen based on whether `V` is passed as an array of indices (Vsparse) or as a dense array of booleans (Vdense).

Here's a makeshift visualization on how the performance of Vsparse vs. Vdense compares when selecting the every *k*th element in the graph. The results are quite noisy: initially, there's no clear winner but the sparse solution is better if <5% of the graph is selected.

![image](https://user-images.githubusercontent.com/1402801/80762495-f3a1fe80-8b3c-11ea-8780-af342ef7361f.png)

I added support for efficiently handling CSC matrices in the Vsparse case:
https://github.com/szarnyasg/LAGraph/blob/fe3db3c0916b2304d20e4b4ea0fccae5cbac5f80/Source/Algorithm/LAGraph_inducedsubgraph.c#L131-L142
